CodenameOne/src/com/codename1/util/DateUtil.java

```diff
@@ -79,7 +79,116 @@ public static int compare(Date d1, Date d2) {
         }
         return 0;
     }
-    
+
+    /**
+     * Compares two dates by the specified field.
+     *
+     * @param d1    The first date to compare
+     * @param d2    The second date to compare
+     * @param field One of the fields:
+     *              <ul>
+     *               <li>{@code Calendar.MILLISECOND}
+     *               <li>{@code Calendar.SECOND}
+     *               <li>{@code Calendar.MINUTE}
+     *               <li>{@code Calendar.HOUR} or  {@code Calendar.HOUR_OF_DAY}
+     *               <li>{@code Calendar.DATE}
+     *               <li>{@code Calendar.MONTH}
+     *               <li>{@code Calendar.YEAR}
+     *
+     * @return <ul>
+     * <li>-1 - if first date is earlier.
+     * <li>1 - if first date is later.
+     * <li>0 - if they are the same.
+     * @since 7.0
+     */
+    public static int compareBy(Date d1, Date d2, int field) {
+        if (d1 == null) return d2 == null ? 0 : -1;
+        if (d2 == null) return 1;
+
+        Calendar c1 = Calendar.getInstance();
+        c1.setTime(d1);
+
+        Calendar c2 = Calendar.getInstance();
+        c2.setTime(d2);
+
+        switch (field) {
```

Comment:
The entire method can be replaced with `return c1.get(d1) - c2.get(d2);`. It should satisfy the comparable contract fine. Comparable doesn't need 1 or -1 for unequal just larger than 0 or smaller than 0 values. If you stick this right where the switch field is and delete everything bewow it work exactly the same as the before.

Reply:
@shai-almog I got what you mean. This should have been the case for the other method `compare(Date d1, Date d2)`.

However, I'm writing this particular method to allow comparison by a specific type of date (day, month, year, hour, minute, or second). If it's just a comparison by millisecond, I wouldn't need the switch statement.

I want to be able to check if `August 2020` is earlier or later than `August 2021`

Reply:
It will work for a field notice I invoke the get() method of calendar. I think this method should return a Comparator interface for Date specifically (notice I can't use Lambda in core code):

```Java
     public static Comparator<Date> compareByDateField(final int field) {
            return new Comparator<Date>() {
                   public int compare(Date object1, Date object2) {
                          Calendar c = Calendar.getInstance();
                          c.setTime(object1);
                          int f1 = c.get(field);
                          c.setTime(object2);
                          int f2 = c.get(field);
                          return f1 - f2;
                   }
            }
     }
```

Now this is useful as you can do things like `Collections.sort` by a specific field etc. which I assume is the actual use case. Right?

Reply:
Using comparator is much better but still won't correctly compare the 2 fields.
Examples:

```
Date date1 = new Date(1572073200000L); // 2019-10-26 9:00:00 CAT
Date date2 = new Date(1601103600000L); // 2020-09-26 9:00:00 CAT
System.out.println(compareByDateField(Calendar.MONTH).compare(date1, date2)); // 1
```
Since `date1` is earlier (even though the month is higher), I expect the result to be less than `0`.

The same thing applies to sort.
```
List<Date> dateList = new ArrayList<>();
dateList.add(new Date(1601103600000L)); // 2020-09-26 9:00:00 CAT
dateList.add(new Date(1569481200000L)); // 2019-09-26 9:00:00 CAT
dateList.add(new Date(1632639600000L)); // 2021-09-26 9:00:00 CAT
dateList.add(new Date(1572073200000L)); // 2019-10-26 9:00:00 CAT

dateList.sort(compareByDateField(Calendar.MONTH));
        
System.out.println(dateList);
```
Output:
```[Sat Sep 26 09:00:00 CAT 2020, Thu Sep 26 09:00:00 CAT 2019, Sun Sep 26 09:00:00 CAT 2021, Sat Oct 26 09:00:00 CAT 2019]```
Expected:
```[Thu Sep 26 09:00:00 CAT 2019, Sat Oct 26 09:00:00 CAT 2019, Sat Sep 26 09:00:00 CAT 2020, Sun Sep 26 09:00:00 CAT 2021]```

Because even though October is later than September, the year October applies to is earlier. Hence I had to reset all fields below `Calendar.MONTH` in the inputs, and then compare the entire calendar objects.

Reply:
To better clear the confusion, I want `October 2019` to be earlier than `September 2020`. when compared by month.

And `1 September 2020 == 30 September 2020` when compared by month.

Though 10 > 9, the year 2019 < 2020.

Reply:
That's because I flipped the return value it should have been:
```java
                          return f2 - f1;
```

Reply:
These two dates shouldn't be equal by comparison using any field since they belong to different years:
```
Date date1 = new Date(1601103600000L); // 2020-09-26 9:00:00 CAT
Date date2 = new Date(1632639600000L); // 2021-09-26 9:00:00 CAT
```
but they are equal except for when compared by year. This is incorrect based on what I'm trying to accomplish.

Reply:
OK, now I understand what you're trying to do. The JavaDocs are really unclear here. You should say that it sorts by the granularity of a specific field. I don't think you need calendar or a switch here. E.g. you can define your own constants for DAY, YEAR etc. where day would be `public static final long DAY = 24 * 60 * 60000;`

Now you can do:

`return d1.getTime() / field - d2.getTime() / field;`

This should work for all cases without a calendar, switch or anything. It should still use a `Comparator`. Does that make sense?

Reply:
It is clear. However, defining those constants may only work for DAY and below. MONTH is inconsistent (28, 29, 30 or 31 days) and YEAR too (365 or 366 days).

Also, there is a possible Numeric Overflow in MONTH and YEAR expressions.

```
public static final long SECOND = 1000;
public static final long MINUTE = 60 * SECOND;
public static final long HOUR = 60 * MINUTE;
public static final long DAY = 24 * HOUR;
public static final long MONTH = 30 * DAY;
public static final long YEAR = 365 * DAY;
```

Reply:
Scrap the above. It will work with below which is using an average of 365.25 days for YEAR and 30.4375 days for MONTH.
So, the chance of inaccuracy is close to zero.
```
public static final long MONTH = 2629800000L;
public static final long YEAR = 31557600000L;
```